File: src/logic/__tests__/emotions.test.ts
import { describe, it, expect } from 'vitest'
import { detectBotEmotions } from '../emotions'
import { Emoticon } from '../../types/emotion'
import { createInitialState, gameReducer } from '../gameReducer'
import { GameActionType, LogEventKey, type GameState } from '../../types/game'
import { actorEntry } from '../logEntries'

function makePlayersState(
  emitterIndex: number,
  opts: { isBot?: boolean; botControlled?: boolean; allBots?: boolean } = {},
): GameState {
  const base = gameReducer(createInitialState(), {
    type: GameActionType.StartGame,
    playerCount: 2,
    names: ['Alice', 'Bob'],
    isBot: [false, true],
  })
  const next = { ...base, eventLog: [] }
  if (opts.allBots) {
    next.players = next.players.map((pl) => ({ ...pl, isBot: true }))
  }
  if (opts.isBot) {
    next.players = next.players.map((pl, i) => (i === emitterIndex ? { ...pl, isBot: true } : pl))
  }
  if (opts.botControlled) {
    next.players = next.players.map((pl, i) =>
      i === emitterIndex ? { ...pl, botControlled: true, isBot: false } : pl,
    )
  }
  return next
}

function withEvent(state: GameState, entries: ReturnType<typeof actorEntry>[]): GameState {
  return { ...state, eventLog: entries }
}

describe('detectBotEmotions', () => {
  it('maps a bot bankruptcy to sad', () => {
    const prev = makePlayersState(1, { isBot: true })
    const bot = prev.players[1]
    const next = withEvent(prev, [actorEntry(LogEventKey.Bankruptcy, bot)])
    expect(detectBotEmotions(prev, next)).toEqual([{ playerId: 1, emoticon: Emoticon.Sad }])
  })

  it('does not emit for a human bankruptcy', () => {
    const prev = makePlayersState(0)
    const human = prev.players[0]
    const next = withEvent(prev, [actorEntry(LogEventKey.Bankruptcy, human)])
    expect(detectBotEmotions(prev, next)).toEqual([])
  })

  it('maps expensive rent (>= threshold) paid by a bot to angry', () => {
    const prev = makePlayersState(1, { isBot: true })
    const bot = prev.players[1]
    const next = withEvent(prev, [
      { key: LogEventKey.PaidRent, params: { name: bot.name, amount: 300, owner: 'Alice' } },
    ])
    expect(detectBotEmotions(prev, next)).toEqual([{ playerId: 1, emoticon: Emoticon.Angry }])
  })

  it('ignores cheap rent below the threshold', () => {
    const prev = makePlayersState(1, { isBot: true })
    const bot = prev.players[1]
    const next = withEvent(prev, [
      { key: LogEventKey.PaidRent, params: { name: bot.name, amount: 299, owner: 'Alice' } },
    ])
    expect(detectBotEmotions(prev, next)).toEqual([])
  })

  it('maps a monopoly rent to proud for the owner bot', () => {
    const prev = makePlayersState(1, { isBot: true })
    const bot = prev.players[1]
    const next = withEvent(prev, [
      { key: LogEventKey.MonopolyRent, params: { owner: bot.name, name: 'Alice' } },
    ])
    expect(detectBotEmotions(prev, next)).toEqual([{ playerId: 1, emoticon: Emoticon.Proud }])
  })

  it('maps a completed trade to happy for each bot party', () => {
    const prev = makePlayersState(1, { isBot: true })
    const bot = prev.players[1]
    const next = withEvent(prev, [
      { key: LogEventKey.TradeAccepted, params: { from: bot.name, to: 'Alice' } },
    ])
    expect(detectBotEmotions(prev, next)).toEqual([{ playerId: 1, emoticon: Emoticon.Happy }])
  })

  it('maps a botControlled (non-isBot) bankruptcy to sad', () => {
    const prev = makePlayersState(0, { botControlled: true })
    const player = prev.players[0]
    const next = withEvent(prev, [actorEntry(LogEventKey.Bankruptcy, player)])
    expect(detectBotEmotions(prev, next)).toEqual([{ playerId: 0, emoticon: Emoticon.Sad }])
  })

  it('maps a completed trade with two bot parties to happy for each bot', () => {
    const prev = makePlayersState(1, { allBots: true })
    const next = withEvent(prev, [
      { key: LogEventKey.TradeAccepted, params: { from: prev.players[0].name, to: prev.players[1].name } },
    ])
    expect(detectBotEmotions(prev, next)).toEqual([
      { playerId: 0, emoticon: Emoticon.Happy },
      { playerId: 1, emoticon: Emoticon.Happy },
    ])
  })

  it('maps doubles to happy for the rolling bot', () => {
    const prev = makePlayersState(1, { isBot: true })
    const bot = prev.players[1]
    const next = withEvent(prev, [actorEntry(LogEventKey.DoublesAgain, bot)])
    expect(detectBotEmotions(prev, next)).toEqual([{ playerId: 1, emoticon: Emoticon.Happy }])
  })

  it('ignores unrelated event-log entries', () => {
    const prev = makePlayersState(1, { isBot: true })
    const bot = prev.players[1]
    const next = withEvent(prev, [
      { key: LogEventKey.Rolled, params: { name: bot.name, d1: 2, d2: 3, total: 5 } },
      { key: LogEventKey.Turn, params: { name: 'Alice' } },
    ])
    expect(detectBotEmotions(prev, next)).toEqual([])
  })
})

File: backend/internal/usecase/botusecase/emotions_test.go
package botusecase

import (
	"reflect"
	"testing"

	"monopoly-game-backend/internal/data"
	"monopoly-game-backend/internal/entity"
)

func emotionTestState() entity.GameState {
	return entity.GameState{
		Players: []entity.Player{
			{ID: 0, Name: "Alice"},
			{ID: 1, Name: "Bob", IsBot: true},
		},
		EventLog: []entity.LogEntry{},
	}
}

func emotionPlayersWith(options func([]entity.Player) []entity.Player) []entity.Player {
	return options(emotionTestState().Players)
}

func TestDetectBotEmotions(t *testing.T) {
	cases := []struct {
		name string
		prev entity.GameState
		next entity.GameState
		want []BotEmotion
	}{
		{
			name: "bot bankruptcy is sad",
			prev: emotionTestState(),
			next: entity.GameState{Players: emotionTestState().Players, EventLog: []entity.LogEntry{{Key: entity.LogEventKeyBankruptcy, Params: map[string]interface{}{"name": "Bob"}}}},
			want: []BotEmotion{{PlayerID: 1, Emoticon: entity.EmoticonSad}},
		},
		{
			name: "human bankruptcy is ignored",
			prev: emotionTestState(),
			next: entity.GameState{Players: emotionTestState().Players, EventLog: []entity.LogEntry{{Key: entity.LogEventKeyBankruptcy, Params: map[string]interface{}{"name": "Alice"}}}},
			want: []BotEmotion{},
		},
		{
			name: "expensive rent is angry",
			prev: emotionTestState(),
			next: entity.GameState{Players: emotionTestState().Players, EventLog: []entity.LogEntry{{Key: entity.LogEventKeyPaidRent, Params: map[string]interface{}{"name": "Bob", "amount": data.StartingMoney / 5}}}},
			want: []BotEmotion{{PlayerID: 1, Emoticon: entity.EmoticonAngry}},
		},
		{
			name: "rent below threshold is ignored",
			prev: emotionTestState(),
			next: entity.GameState{Players: emotionTestState().Players, EventLog: []entity.LogEntry{{Key: entity.LogEventKeyPaidRent, Params: map[string]interface{}{"name": "Bob", "amount": entity.EXPENSIVE_RENT_THRESHOLD - 1}}}},
			want: []BotEmotion{},
		},
		{
			name: "monopoly rent is proud",
			prev: emotionTestState(),
			next: entity.GameState{Players: emotionTestState().Players, EventLog: []entity.LogEntry{{Key: entity.LogEventKeyMonopolyRent, Params: map[string]interface{}{"owner": "Bob"}}}},
			want: []BotEmotion{{PlayerID: 1, Emoticon: entity.EmoticonProud}},
		},
		{
			name: "accepted trade makes each bot party happy",
			prev: emotionTestState(),
			next: entity.GameState{Players: emotionPlayersWith(func(players []entity.Player) []entity.Player { players[0].IsBot = true; return players }), EventLog: []entity.LogEntry{{Key: entity.LogEventKeyTradeAccepted, Params: map[string]interface{}{"from": "Alice", "to": "Bob"}}}},
			want: []BotEmotion{{PlayerID: 0, Emoticon: entity.EmoticonHappy}, {PlayerID: 1, Emoticon: entity.EmoticonHappy}},
		},
		{
			name: "bot controlled bankruptcy is sad",
			prev: emotionTestState(),
			next: entity.GameState{Players: emotionPlayersWith(func(players []entity.Player) []entity.Player { players[0].BotControlled = true; return players }), EventLog: []entity.LogEntry{{Key: entity.LogEventKeyBankruptcy, Params: map[string]interface{}{"name": "Alice"}}}},
			want: []BotEmotion{{PlayerID: 0, Emoticon: entity.EmoticonSad}},
		},
		{
			name: "doubles makes bot happy",
			prev: emotionTestState(),
			next: entity.GameState{Players: emotionTestState().Players, EventLog: []entity.LogEntry{{Key: entity.LogEventKeyDoublesAgain, Params: map[string]interface{}{"name": "Bob"}}}},
			want: []BotEmotion{{PlayerID: 1, Emoticon: entity.EmoticonHappy}},
		},
		{
			name: "unrelated events are ignored",
			prev: emotionTestState(),
			next: entity.GameState{Players: emotionTestState().Players, EventLog: []entity.LogEntry{{Key: entity.LogEventKeyRolled, Params: map[string]interface{}{"name": "Bob"}}, {Key: entity.LogEventKeyTurn, Params: map[string]interface{}{"name": "Alice"}}}},
			want: []BotEmotion{},
		},
	}

	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			if got := DetectBotEmotions(tc.prev, tc.next); !reflect.DeepEqual(got, tc.want) {
				t.Fatalf("DetectBotEmotions() = %#v, want %#v", got, tc.want)
			}
		})
	}
}

func TestDetectBotEmotionsUsesOnlyNewEntriesAndHandlesMalformedParams(t *testing.T) {
	prev := emotionTestState()
	prev.EventLog = []entity.LogEntry{{Key: entity.LogEventKeyBankruptcy, Params: map[string]interface{}{"name": "Bob"}}}
	next := prev
	next.EventLog = append(append([]entity.LogEntry{}, prev.EventLog...), entity.LogEntry{Key: entity.LogEventKeyPaidRent, Params: map[string]interface{}{"name": "Bob", "amount": "300"}})
	if got := DetectBotEmotions(prev, next); !reflect.DeepEqual(got, []BotEmotion{}) {
		t.Fatalf("malformed or old entries produced emotions: %#v", got)
	}
}
